Display the link to add the shipping address when shipping address is not available

tarunvijwani opened the pull request:
With this PR, we display a link to add the shipping address in the shipping calculator on the cart page when the shipping address is unavailable. We're also hiding the option to change the shipping address when default shipping rates are defined. 

Fixes #8027 and  #7730

### Changes in the PR
- Replaced hardcoded notices with text for Cart and Checkout block.
- Fixed the shipping placeholder for the Checkout block when no shipping rates are available.
- Created a function in base/utils to format Shipping address.
- Removed the condition to check the default customer address. Display the "Add shipping address" link only when no address is available.
- Hide shipping calculator if default shipping is defined.


#### Accessibility

<!-- If you've changed or added any interactions, check off the appropriate items below. You can delete any that don't apply. Use this space to elaborate on anything if needed. -->

- [x] I've tested using only a keyboard (no mouse)
- [x] I've tested using a screen reader
- [ ] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [ ] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

### Screenshots

<!-- If your change has a visual component, add a screenshot here. A "before" screenshot would also be helpful. -->

| Before | After |
| ------ | ----- |
| <img width="100%" alt="image" src="https://user-images.githubusercontent.com/11503784/211488531-ba2e119e-e1e3-4f2a-9852-d8e4f0b1dc37.png">  |<img width="100%" alt="image" src="https://user-images.githubusercontent.com/11503784/226824722-b92e00d0-faba-4a74-b981-5fcd722867ca.png">|

### Testing

#### Automated Tests
* [x] Changes in this PR are covered by Automated Tests.
  * [x] Unit tests
  * [ ] E2E tests

#### User Facing Testing

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->
### case 1 (With no default shipping rates)

<img width="1429" alt="image" src="https://user-images.githubusercontent.com/11503784/214127269-1cb4d478-88b7-4343-b841-96c35300fd6d.png">

1. Check out this branch.
2. Go to the WooCommerce settings page: `wp-admin/admin.php?page=wc-settings`.
3. Change 'Default customer location' to `No location by default`. 
4. Remove any default(fallback) shipping rates. 
5. Open your site in a private session(incognito mode).
6. Add some products to the cart. 
7. Go to the cart page, and confirm the "Add an address for shipping rates" link is present in the shipping calculator.
8. Click on the link, and confirm it is working as expected.
9. Open your site in a regular session where you're already logged in and have placed an order before. 
10. Add some products to the cart. 
11. Go to the cart page, confirm the shipping address, and "change address" link present in the shipping calculator.
12. Click on the link, and confirm it is working as expected.

### case 2(With default shipping rates)

<img width="1286" alt="image" src="https://user-images.githubusercontent.com/11503784/214127480-f8387749-d45b-41fe-b0d5-db29665eb759.png">

1. Check out this branch.
2. Go to the WooCommerce settings page: `wp-admin/admin.php?page=wc-settings`.
3. Change 'Default customer location' to `No location by default`. 
4. Add any default(fallback) shipping rates. 
5. Open your site in a private session(incognito mode).
6. Add some products to the cart. 
7. Go to the cart page, and confirm the shipping calculator is hidden and default shipping rates are visible
8. Open your site in a regular session where you're already logged in and have placed an order before. 
9. Add some products to the cart. 
10. Go to the cart page, confirm the shipping address, and "change address" link present in the shipping calculator.
11. Click on the link, and confirm it is working as expected.

### case 3(Error Notice)
1. Check out this branch.
2. Go to the WooCommerce settings page: `wp-admin/admin.php?page=wc-settings`.
3. Change 'Default customer location' to `No location by default`. 
4. Remove any default(fallback) shipping rates. 
5. Open your site in a private session(incognito mode).
6. Add some products to the cart. 
7. Go to the cart page, and confirm the "Add an address for shipping rates" link is present in the shipping calculator.
8. Click on the link, and change the address to a location where shipping rates are unavailable.
9. Confirm that `There are no shipping options available. Please check your shipping address.` error is visible in the text form.
10. Proceed to the Checkout block page, and confirm the `There are no shipping options available. Please check your shipping address` error is visible in the text form under Shipping options. 

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental



### Changelog

> Display the link to add the shipping address when the shipping address is not available.

mikejolley commented:
Before looking at the code, is it not strange that the text here states "Add an address for shipping rates", but still displays shipping rates below? It seems like this link should only change if there are no rates displayed. Unless of course the link text is made slightly more generic so it fits both cases.

tarunvijwani commented:
>Before looking at the code, is it not strange that the text here states "Add an address for shipping rates", but still displays shipping rates below? It seems like this link should only change if there are no rates displayed. Unless of course the link text is made slightly more generic so it fits both cases.

That's a great point. I think we can change the text to something like "Add an address to get the exact shipping rates"

@elizaan36 @ralucaStan I'd love to hear your thoughts.

mikejolley commented on assets/js/base/components/cart-checkout/totals/shipping/index.tsx:
```diff
@@ -60,15 +60,28 @@ const ShippingAddress = ( {
 	setIsShippingCalculatorOpen,
 	shippingAddress,
 }: ShippingAddressProps ): ReactElement | null => {
+	const hasDefaultCustomerAddress = getSetting(
```
I notice that in the cart block, if you have an address (whether you've entered it or not) the `ShippingLocation` component is passed a `shippingAddress`. The check inside that component generates a formatted address and returns if null:

```
if ( ! formattedLocation ) {
		return null;
	}
```

Could we instead move that code to a util and consume in both places?

tarunvijwani replied:
That's a great suggestion. Thank you!  I have updated the code. 👍

mikejolley commented on src/BlockTypes/Cart.php:
```diff
@@ -194,6 +194,7 @@ function() {
 		$this->asset_data_registry->add( 'isShippingCalculatorEnabled', filter_var( get_option( 'woocommerce_enable_shipping_calc' ), FILTER_VALIDATE_BOOLEAN ), true );
 		$this->asset_data_registry->add( 'displayItemizedTaxes', 'itemized' === get_option( 'woocommerce_tax_total_display' ), true );
 		$this->asset_data_registry->add( 'displayCartPricesIncludingTax', 'incl' === get_option( 'woocommerce_tax_display_cart' ), true );
+		$this->asset_data_registry->add( 'hasDefaultCustomerAddress', ! empty( get_option( 'woocommerce_default_customer_address' ) ), true );
```
👍🏻

mikejolley reviewed (changes requested):
The logic makes sense, just a question on reuse and then perhaps a change to the text if needed.

tarunvijwani commented on assets/js/base/components/cart-checkout/totals/shipping/index.tsx:
```diff
@@ -60,15 +61,25 @@ const ShippingAddress = ( {
+
+	const formattedLocation = formatShippingAddress( shippingAddress );
+	const label =
+		! hasDefaultCustomerAddress && ! formattedLocation
```
@mikejolley Do you think we should remove `hasDefaultCustomerAddress` check as well? I think `formatLocation` null check is enough to determine whether we have the address or not to change the label, irrespective of the user's setting in WC. Let me know your thoughts.

mikejolley replied:
Yeah that makes sense, I'm assuming that formatted location logic needs the basics (state, country, city) or will null. If thats the case, yes remove it.

elizaan36 commented:
hey @tarunvijwani @mikejolley 

It's my understanding that the text "Add an address for shipping rates" would only appear when there are no rates shown. What's the use case where the link would show alongside rates?

tarunvijwani commented:
> It's my understanding that the text "Add an address for shipping rates" would only appear when there are no rates shown. What's the use case where the link would show alongside rates?

@elizaan36 If the merchant has defined the default shipping rates that are applicable for every location, then it will show the rates.  If the customer adds their address, the shipping rates will change according to their location. The Cart Shortcode page also follows the same approach. 


https://user-images.githubusercontent.com/11503784/213396368-416cf45b-c9e2-4585-afbd-21d1c22e4773.mov





However, if the merchant has not defined the default shipping rates, then it will display the error(We're adding the changes of #8130 to this PR to keep the text consistent) 


https://user-images.githubusercontent.com/11503784/213396404-a9208794-56e8-4db5-8625-7fe1b83c8afc.mov

elizaan36 commented:
> @elizaan36 If the merchant has defined the default shipping rates that are applicable for every location, then it will show the rates.

I imagine the merchant's intention would be to offer the flat rate and turn off the shipping calculator in that case. So I'd say if the merchant has chosen a flat rate for all locations, the link shouldn't appear.

It's quite confusing for the shopper to see "Flat rate" or "Free shipping" with a link that might change these rates. What do you think?

> However, if the merchant has not defined the default shipping rates, then it will display the error(We're adding the changes of https://github.com/woocommerce/woocommerce-blocks/pull/8130 to this PR to keep the text consistent)

Nice, this makes sense and glad we can align the experience there!

tarunvijwani commented:
>I imagine the merchant's intention would be to offer the flat rate and turn off the shipping calculator in that case. So I'd say if the merchant has chosen a flat rate for all locations, the link shouldn't appear.
>It's quite confusing for the shopper to see "Flat rate" or "Free shipping" with a link that might change these rates. What do you think?

Yes, that makes perfect sense! I'll incorporate the changes. Thank you!

Commit:
Remove hardcoded notices & change display shipping calculator codition

- Replace hardcoded notices with text for Cart and Checkout block.
- Fix shipping placeholder for the Chec

Commit:
Remove hardcoded notices & change display shipping calculator condition

- Replace hardcoded notices with text for Cart and Checkout block.
- Fix shipping placeholder for the Checkout block when no shipping rates are available.
- Created a function in base/utils to format Shipping address.
- Remove the condition to check default customer address. Display the "Add shipping address" link only when there is no address available.
- Hide shipping calculator if default shipping  are defined.

mikejolley commented on assets/js/base/components/cart-checkout/shipping-location/index.tsx:
```diff
@@ -2,56 +2,21 @@
- * @param {Object} props         Incoming props for the component.
- * @param {Object} props.address Incoming address information.
+ * @param {Object} props                   Incoming props for the component.
+ * @param {string} props.formattedLocation Formatted location.
```
You could move this comment to the interface above and remove the docblock comments here.

mikejolley commented on assets/js/base/components/cart-checkout/totals/shipping/shipping-placeholder.tsx:
```diff
@@ -39,6 +39,10 @@ export const ShippingPlaceholder = ( {
 	return (
 		<CalculatorButton
+			label={ __(
+				'Add an address for shipping rates',
```
This uses the term "shipping rates", yet the message below and in other places uses "shipping options". We should be consistent because they both mean the same thing.

tarunvijwani replied:
Right, we should keep the text the same everywhere. I have updated the label. Thank you! 👍

mikejolley commented on assets/js/base/utils/address.ts:
```diff
@@ -100,3 +102,53 @@ export const emptyHiddenAddressFields = <
+ * @param {Object} address The address to format.
+ * @return {string | null} The formatted address or null if no address is provided.
+ */
+export const formatShippingAddress = (
```
👍🏻

mikejolley commented on assets/js/blocks/checkout/inner-blocks/checkout-order-summary-shipping/block.tsx:
```diff
@@ -26,6 +26,7 @@ const Block = ( {
 				showRateSelector={ false }
 				values={ cartTotals }
 				currency={ totalsCurrency }
+				isCheckout={ true }
```
What was this for?

tarunvijwani replied:
I found a bug on the checkout block page that displays `Calculated during checkout` for shipping in the Checkout summary if shipping options are unavailable. It happened because we were not passing the `isCheckout` prop to [TotalsShipping component](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/blocks/checkout/inner-blocks/checkout-order-summary-shipping/block.tsx#L24) for Checkout.  This change will fix it and display the correct error message.
![image](https://user-images.githubusercontent.com/11503784/215052665-17b8f1ad-7b33-43be-a689-cf2c356d345a.png)

mikejolley reviewed (changes requested):
Something is broken here. I just tested with no default customer location in a new session, and this is what I got:

![Screenshot 2023-01-24 at 11 40 14](https://user-images.githubusercontent.com/90977/214282587-18694015-81c6-4f0d-aa94-e3319ee79a8f.png)

No location is shown, and no shipping calculator is shown.

opr commented on assets/js/base/components/cart-checkout/shipping-location/index.tsx:
```diff
@@ -2,56 +2,21 @@
- * @param {Object} props.address Incoming address information.
+ * @param {Object} props                   Incoming props for the component.
+ * @param {string} props.formattedLocation Formatted location.
+ * @return {JSX.Element|null} The component.
```
We could remove this, the TS type is good enough I think.

opr commented on assets/js/base/utils/address.ts:
```diff
@@ -100,3 +102,53 @@ export const emptyHiddenAddressFields = <
+	if ( Object.values( address ).length === 0 ) {
+		return null;
+	}
+	const shippingCountries = getSetting( 'shippingCountries', {} ) as Record<
+		string,
+		string
+	>;
```
instead of using `as Record< string, string >` you can pass the generic type to `getSetting` like so:

```suggestion
	const shippingCountries = getSetting< Record< string, string > >(
		'shippingCountries',
		{}
	);
```

This can apply to the other instances of this too :)

tarunvijwani replied:
Thank you for the suggestion. I have updated the code.  👍

opr commented on assets/js/base/components/cart-checkout/totals/shipping/shipping-address.tsx:
```diff
@@ -23,9 +24,13 @@ export const ShippingAddress = ( {
 	return (
 		<>
-			<ShippingLocation address={ shippingAddress } />
+			<ShippingLocation formattedLocation={ formattedLocation } />
```
Nice change, this will be helpful for #7997

opr reviewed (changes requested):
Hi Tarun, thanks for working on this.

It's looking good, but there are a few things that should be looked at:

I see the 'Add an address for shipping rates' displayed inline, I don't see this in the designs, is this intended? I don't think it looks bad, just want to double check.

<img width="372" alt="image" src="https://user-images.githubusercontent.com/5656702/214328872-22e6bb26-176f-4f23-a841-ca22b5e2714e.png">

A couple of TS changes, and I won't repeat what Mike already mentioned.

Commit:
Merge branch 'trunk' into add/8027-shipping-address-link

tarunvijwani commented:
> Something is broken here. I just tested with no default customer location in a new session, and this is what I got:
> 
> ![Screenshot 2023-01-24 at 11 40 14](https://user-images.githubusercontent.com/90977/214282587-18694015-81c6-4f0d-aa94-e3319ee79a8f.png)
> 
> No location is shown, and no shipping calculator is shown.

>

@mikejolley do you have default shipping rates enabled? If yes, this is expected as per Test case 2(With default shipping rates). We [decided to remove the shipping calculator](https://github.com/woocommerce/woocommerce-blocks/pull/8141#issuecomment-1397239622) if default shipping rates are defined.

tarunvijwani commented:
> I see the 'Add an address for shipping rates' displayed inline, I don't see this in the designs, is this intended? I don't think it looks bad, just want to double check.

@opr This might be happening because of the existing CSS for the calculate button component. I changed the text as per Mike's suggestion. Since the text is long so, it appears in the next line now. However, it will vary from theme to theme. I think it is fine to keep the text as is. However, if you feel that we should change the style of the component, let me know. 
![image](https://user-images.githubusercontent.com/11503784/215055977-a6a8eb49-9d40-435c-910d-64380edc6d38.png)

opr commented:
Hey @tarunvijwani thanks for explaining, maybe @elizaan36 could help clarify whether this change is ok or if it should be on the line below at all times.

Commit:
Fix the label of the calculator button and TypeScript

- Change the label of the calculator button from "Add an address for shipping rates" to  "Add an address for shipping options".
- Fix the TypeScript of the utils/address file.
- Remove the docblock comments from shipping location block

Commit:
Fix the label of the calculator button and TypeScript

- Change the label of the calculator button from "Add an address for shipping rates" to  "Add an address for shipping options".
- Fix the TypeScript of the utils/address file.
- Remove the docblock comments from shipping location block

tarunvijwani commented:
Translation for the text: `Add an address for shipping options` is [not available at the moment](https://translate.wordpress.org/projects/wp-plugins/woo-gutenberg-products-block/stable/nl/default/?filters%5Bterm%5D=Add+an+address+for+shipping+options&filters%5Bterm_scope%5D=scope_any&filters%5Bstatus%5D=current_or_waiting_or_fuzzy_or_untranslated&filters%5Buser_login%5D=&filter=Apply+Filters&sort%5Bby%5D=priority&sort%5Bhow%5D=desc). If the PR is approved, I'll wait till Feb 3 to merge the PR to trunk so that it will be picked in the next WC core release.

elizaan36 commented:
hey thanks for the ping here, Ideally the "add address" link should appear on it's own line (that was the intention) but it's not a huge UX issue if it appears inline in some cases. 

Looking at the screenshot now I'm mainly curious why it's not picking up the same text links styles as the rest of the page? The text is gray and heavier weight, which creates a lot of visual noise. It should pull in the accent color defined by the theme for text links and match the font weight.

<img width="367" alt="image" src="https://user-images.githubusercontent.com/5121465/215186901-154bb07f-c9cb-4ddb-92f5-6ecf3d3f1bad.png">

Commit:
Merge branch 'trunk' into add/8027-shipping-address-link

Commit:
Change the link style and fix editor issue

- If no shipping rates are available, we will display a link instead of a button.
- Display change address button when we're in Editor as some e2e tests are dependent on it

Commit:
Change the link style and fix editor issue

- If no shipping rates are available, we will display a link instead of a button.
- Display change address button when we're in Editor as some e2e tests are dependent on it

tarunvijwani commented:
@elizaan36 Fixed it in the [commit](https://github.com/woocommerce/woocommerce-blocks/pull/8141/commits/0b0d6186cb874acc7a039a1eb05aaf042acb5d90):

<img width="853" alt="image" src="https://user-images.githubusercontent.com/11503784/215512799-e431135d-7c80-4d4e-aed7-ecbd18e304ac.png">

opr commented on assets/js/base/components/cart-checkout/totals/shipping/calculator-button.tsx:
```diff
@@ -5,25 +5,48 @@ import { __ } from '@wordpress/i18n';
-			className="wc-block-components-totals-shipping__change-address-button"
+		<a
+			role="button"
+			href="#wc-block-components-shipping-calculator-address__link"
```
This causes the URL in the browser to update when clicking. It is unclear to me why we have to show a link instead of a button when rates are not available, what is the specific reason for showing different elements?

To avoid the URL change you can pass the event as a parameter of the `onClick` handler, and call `.preventDefault()` on it, like so:

```js
onClick={ ( e ) => {
  e.preventDefault();
  setIsShippingCalculatorOpen( ! isShippingCalculatorOpen );
} }
```

tarunvijwani replied:
Ah, I missed adding the ` e.preventDefault();`. Thank you, I have updated the code.  

>It is unclear to me why we have to show a link instead of a button when rates are not available, what is the specific reason for showing different elements?

It's because of the styling of the elements. I kept the original button for the conditions where we are displaying the `calculate` link. 
For others, I'm displaying the `Add an address` link to match the style with the coupon code link. 

I tried using the link at both places but then the `calculate` link gets the theme color which was taking too much attention. So I decided to keep the original style and elements for it. Please let me know if there is a better way to tackle it.

opr reviewed (commented):
Hey Tarun, when I add an address that does not have rates I see this:

<img width="317" alt="image" src="https://user-images.githubusercontent.com/5656702/215759883-8f05d11c-abe3-486e-8a07-95c76687831d.png">

I wonder if it would be better if we checked to see if the address is not empty, and if so, we could change the text to `Add a different address for shipping options`. The reason this may be better is because the shopper could be confused as they would have just added an address, so why would the page be asking them to add one? Also added a comment about the link/button situation, it would be great if you could provide more info on that! 🙇🏼

Commit:
Prevent default for the add address link

Add the prevent default for the add address link so that the URL in the browser won't get updated.

Commit:
Prevent default for the add address link

Add the prevent default for the add address link so that the URL in the browser won't get updated.

tarunvijwani commented:
> I wonder if it would be better if we checked to see if the address is not empty, and if so, we could change the text to Add a different address for shipping options. The reason this may be better is because the shopper could be confused as they would have just added an address, so why would the page be asking them to add one? Also added a comment about the link/button situation, it would be great if you could provide more info on that! 🙇🏼

That's a great suggestion. However, if we change the text, it will be misaligned for some themes:

![image](https://user-images.githubusercontent.com/11503784/216737147-58057dd7-6a82-431e-b78e-749d6628e16c.png)


I tried adding the `white-space: break-spaces;` in CSS but it is looking ugly. 


We can try a shorter length string like `Add a different shipping address`. Let me know your thoughts.

opr reviewed (commented):
Ok, thanks for the answer, if we want to keep it in line with the coupon link this is OK 👍🏼 

As for the text not breaking, this is caused by https://github.com/woocommerce/woocommerce-blocks/blob/669aee78ad4b6c97388b2e75e444ebb38df23be5/packages/checkout/components/totals/item/style.scss#L14 so maybe we can apply a new class, or modifier or something and remove this specific rule for the add/update shipping address link?

elizaan36 commented:
> I wonder if it would be better if we checked to see if the address is not empty, and if so, we could change the text to Add a different address for shipping options. The reason this may be better is because the shopper could be confused as they would have just added an address, so why would the page be asking them to add one?

@tarunvijwani @opr Would it be possible to align with the behavior on the checkout block for this use case. 
- User enters an address that is not within any shipping zone
- Notice appears "There are no shipping options available. Please check your shipping address."

elizaan36 commented:
Something like this - thoughts?
<img width="632" alt="image" src="https://user-images.githubusercontent.com/5121465/219533318-748a3971-d13f-42b7-abbe-608c00b8c557.png">

Commit:
Add error notice when no shipping rates available for the given address

- Add error notice when no shipping rates available for the given address.
- Remove conditions display button instead of the link for shipping calculator.
- Replace calculator button with link to match "Add a coupon" link style.

Commit:
Merge branch 'trunk' into add/8027-shipping-address-link

Commit:
Merge remote-tracking branch 'origin/add/8027-shipping-address-link' into add/8027-shipping-address-link

Commit:
Remove hasRates prop from shipping placeholder

Commit:
Replace Notice with text message in Cart

- Replace Notice with text message in Cart.
- Change the variable name from addressComplete to isAddressComplete to keep it consistent.

Commit:
Merge branch 'trunk' into add/8027-shipping-address-link

tarunvijwani commented:
@elizaan36 Just wanted to share the final output of the PR. Please feel free to add any changes and suggestions. Thank you! 

## Case 1: With no default shipping method:

https://user-images.githubusercontent.com/11503784/220379739-8f634652-0fbd-4dbf-bd6c-a329778c58b0.mov

## Case 2: With default shipping method:

https://user-images.githubusercontent.com/11503784/220380918-5f7a9ab6-2da4-4835-bfc6-9d642b0519f3.mov

elizaan36 commented:
Thanks for sharing @tarunvijwani ! One small thing. To help with visual rhythm and scanning the page, can we match the padding top and bottom around the rates? Here's a sketch to illustrate. 
<img width="317" alt="Screen Shot 2023-02-21 at 9 25 02 AM" src="https://user-images.githubusercontent.com/5121465/220388187-ad2e7c24-8c58-417a-af49-e4580f0bdbe6.png">

The checkout block has a good amount of padding around the rates, so we should aim for a cohesive look and feel on the Cart.

Commit:
Add margin-top to shipping rate selector on Cart block

tarunvijwani commented:
> One small thing. To help with visual rhythm and scanning the page, can we match the padding top and bottom around the rates? Here's a sketch to illustrate.

Thank you for bringing it up. I have updated the margin now:

| Twenty Twenty-Three | Storefront |
|---|---|
|![image](https://user-images.githubusercontent.com/11503784/220405647-6f7b4da5-5e9a-4527-ac86-e4e4bf3fbcf7.png)|![image](https://user-images.githubusercontent.com/11503784/220405188-8e7b445c-3ff6-4f42-b0ac-ffaee0d6d480.png)|

ralucaStan commented:
Noting here as well that we don't have the informational notices (that @elizaan36 used in the design) implemented yet in the project.

So we can either use the text variant from Tarun's videos or we can show the error notice and follow up with some other tickets in the future to replace it with the information one.

tarunvijwani commented:
Converted the PR to draft to add unit test cases.

Commit:
Add unit cases for isAddressComplete and formatShippingAddress functions

- Add isAddressComplete as formatShippingAddress was returning false for the empty address.
- Replace formatShippingAddress function with isAddressComplete to validate the address.
- Add unit cases for isAddressComplete and formatShippingAddress functions

tarunvijwani commented:
@opr While creating the unit test cases, I found an error that `formatShippingAddress` returns the formatted address string even if the postcode or any other field is missing.  I have created the `isAddressComplete` function to validate the address and replaced the `formatShippingAddress` function with it. Please let me know if you have any questions.

Commit:
Merge branch 'trunk' into add/8027-shipping-address-link

opr commented on assets/js/base/utils/test/address.ts:
```diff
@@ -22,3 +26,101 @@ describe( 'emptyHiddenAddressFields', () => {
+		expect( isAddressComplete( address ) ).toBe( true );
+	} );
+
+	it( 'correctly checks is  complete', () => {
```
```suggestion
	it( 'correctly checks complete addresses', () => {
```

opr commented on assets/js/base/utils/test/address.ts:
```diff
@@ -22,3 +26,101 @@ describe( 'emptyHiddenAddressFields', () => {
+		expect( isAddressComplete( address ) ).toBe( false );
+	} );
+
+	it( 'correctly checks is not complete', () => {
```
```suggestion
	it( 'correctly checks incomplete addresses', () => {
```

opr commented on assets/js/base/utils/test/address.ts:
```diff
@@ -22,3 +26,101 @@ describe( 'emptyHiddenAddressFields', () => {
 } );
+
+describe( 'isAddressComplete', () => {
+	it( 'correctly checks address is empty', () => {
```
```suggestion
	it( 'correctly checks empty addresses', () => {
```

opr commented on assets/js/base/utils/test/address.ts:
```diff
@@ -22,3 +26,101 @@ describe( 'emptyHiddenAddressFields', () => {
+} );
+
+describe( 'formatShippingAddress', () => {
+	it( 'correctly returns null address is empty', () => {
```
```suggestion
	it( 'returns null if address is empty', () => {
```

opr commented on assets/js/base/components/cart-checkout/totals/shipping/shipping-address.tsx:
```diff
@@ -23,13 +28,21 @@ export const ShippingAddress = ( {
+	const addressComplete = isAddressComplete( shippingAddress );
+	const { isEditor } = useEditorContext();
+
+	// If there is no address, and we're not in the editor, don't show anything.
```
```suggestion
	// If the address is incomplete, and we're not in the editor, don't show anything.
```

opr commented on assets/js/base/utils/address.ts:
```diff
@@ -100,3 +102,62 @@ export const emptyHiddenAddressFields = <
+};
+
+/**
+ * Returns true if the address have city, country and postcode.
```
```suggestion
 * Returns true if the address has a city, country, and postcode.
```

opr commented on assets/js/base/utils/address.ts:
```diff
@@ -100,3 +102,62 @@ export const emptyHiddenAddressFields = <
+export const formatShippingAddress = (
+	address: ShippingAddress | BillingAddress
+): string | null => {
+	// we bail early if we don't have an address.
```
Nitpick, sorry, but comments should begin with a capital letter.

```suggestion
	// We bail early if we don't have an address.
```

opr reviewed (changes requested):
The new tests look nice, I suggested some improvements to the names to make them a little more accurate. By the way there is a typo in the email and customer name, it should be `John Doe` (this isn't technically important as the tests still work, but it is just a small detail that makes our repo appear more polished).

Besides this, I still really think we should change the wording when an address has _already_ been entered.

<img width="384" alt="image" src="https://user-images.githubusercontent.com/5656702/221277740-3ef55eb6-804c-4fa7-9dc8-666fdf93f3c6.png">

If the user is being asked to add an address, it implies one has not been entered. We can use your `isAddressComplete` function combined with the number of rates to determine what to show in this case.

If this presents too many problems, let's resolve the current suggestions, and we can merge this and address the link wording in a follow-up PR, how does that sound?

Commit:
Update assets/js/base/utils/test/address.ts

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

Commit:
Update assets/js/base/components/cart-checkout/totals/shipping/shipping-address.tsx

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

Commit:
Update assets/js/base/utils/test/address.ts

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

Commit:
Update assets/js/base/utils/address.ts

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

Commit:
Update assets/js/base/utils/address.ts

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

Commit:
Update assets/js/base/utils/test/address.ts

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

Commit:
Update assets/js/base/utils/test/address.ts

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

Commit:
Fix typo in unit test data

- Fix typo in unit test data by changing the customer name from "John Deo" to " John Doe".
- Fix typo in surname for email in unit test data.

Commit:
Remove postcode from isAddressComplete check

- Some countries like UAE doesn't have post code. The isAddressComplete function was returning false for these countries. Remove postcode check from isAddressComplete function to make sure it works for all the countries. We already have the postcode validation in shipping calculator that works for the countries which requires postcode.

tarunvijwani commented on assets/js/base/utils/address.ts:
```diff
@@ -100,3 +102,62 @@ export const emptyHiddenAddressFields = <
+export const isAddressComplete = (
+	address: ShippingAddress | BillingAddress
+): boolean => {
+	return !! address.city && !! address.country;
```
Some countries like UAE don't have postcodes. The `isAddressComplete` function was returning `false` for these countries. I have now removed the postcode check from the `isAddressComplete` function to ensure it works for all countries. We already have the postcode validation in the shipping calculator that works for the countries which require postcodes.

opr reviewed (approved):
Hey @tarunvijwani nice it looks good, sorry I missed the change with the address shown!

tarunvijwani commented:
We have open #8621 PR to add more unit tests to this one. Once #8621 is merged, we will test this PR again and merge it.

Commit:
Add unit tests for shipping rate calculator (#8621)

* Add further unit tests

* Add useStoreCart to mocked values and wrap in SlotFillProvider

* Add test for 2nd and 3rd scenario

* Remove AsyncModeProvider mock

Remove AsyncModeProvider mock since we are requiring @wordpress/data in setup-globals.js

---------

Co-authored-by: Thomas Roberts <thomas.roberts@automattic.com>

Commit:
Merge branch 'trunk' into add/8027-shipping-address-link

tarunvijwani commented:
This PR will be merged in the 1st week of April and released in the 10.0.0 version.

Commit:
Merge branch 'trunk' into add/8027-shipping-address-link

Commit:
Fix linting error

Fix linting error that got introduced while resolving the merge conflict.

Commit:
Display correct message in checkout when address shipping is complete

- Rename the isShippingAddressComplete variable to shippingAddressHasErrors as it only checks validation errors.
- Use isAddressComplete function to check if address is complete.

Commit:
Merge branch 'trunk' into add/8027-shipping-address-link

tarhi-saad commented on assets/js/base/components/cart-checkout/totals/shipping/index.tsx:
```diff
@@ -77,23 +80,26 @@ export const TotalsShipping = ( {
+						? totalShippingValue
+						: // if address is not complete, display the link to add an address.
+						  ! addressComplete && (
+								<ShippingPlaceholder
```
@tarunvijwani, on the Checkout Block page, we would like to show the "ShippingPlaceholder" component when there are no shipping methods available (i.e., `hasRates && cartHasCalculatedShipping`). But, with this additional condition (i.e., `! addressComplete &&`), the Shipping placeholder will never be displayed if the address is complete!

tarunvijwani replied:
Continuing the discussion on slack: p1679930401191499-slack-C8X6Q7XQU

tarhi-saad reviewed (commented):
@tarunvijwani, there is a logical error in the code! Please check my comment below!
